### PR TITLE
users: add new resource

### PR DESCRIPTION
**Pull request by BadrAly**
users: add new resource

* Adds a new 'user' resource to manage the user profiles. It contains
  the user personal data as opposed to the patron data that are link to
  a specific organisation.
* Writes a REST API:
  * A GET, POST, PUT methods to create, update and retrieves the patron
    personal information.
  * A search endpoint to retrieve the patron personal information given
    an email or a username.
* Creates units testing.
* Centralizes several views and methods in a specific `User` python
  class.
* Disables the validation email for freshly created user.
* Uses the standard ngx-formly `widget` field for the user JSON schema.

Co-Authored-by: Aly Badr <aly.badr@rero.ch>
Co-Authored-by: Johnny Mariéthoz <Johnny.Mariethoz@rero.ch>

## Code review check list

- [ ] Commit message template compliance.
- [ ] Commit message without typos.
- [ ] File names.
- [ ] Functions names.
- [ ] Functions docstrings.
- [ ] Unnecessary commited files?
- [ ] Cypress tests successful?

### CODE REVIEW

File: rero_ils/modules/users/api.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+            'links': {'self': url_for(
+                'api_users.users_item', _external=True, id=self.user.id)},
+            'metadata': {}
+        }
+        for field in self.profile_fields:
+            value = getattr(self.user.profile, field)
+            if field == 'birth_date':
+                value = datetime.strftime(value, '%Y-%m-%d')
+            data['metadata'][field] = value
+        data['metadata'][field] = self.user.email
```
Comment (rerowep) [inline rero_ils/modules/users/api.py:145]: Schould it be `data['metadata']['email'] = self.user.email`

File: rero_ils/modules/users/api.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+                send_reset_password_instructions(user)
+            db.session.merge(user)
+        db.session.commit()
+        return self
+
+    @classmethod
+    def _validate(cls, data, **kwargs):
+        """Validate user record against schema."""
+        format_checker = FormatChecker()
+        schema = data.pop('$schema', None)
```
Comment (rerowep) [inline rero_ils/modules/users/api.py:112]: Does the validation function correctly? In the schema file the `$schema` is a required field?

File: rero_ils/modules/users/jsonschemas/users/user-v0.0.1.json [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    "last_name",
+    "birth_date",
+    "username"
+  ],
+  "properties": {
+    "$schema": {
+      "title": "Schema",
+      "description": "Schema to validate user records against.",
+      "type": "string",
+      "minLength": 9,
```
Comment (rerowep) [inline rero_ils/modules/users/jsonschemas/users/user-v0.0.1.json:30]: Pattern validation should be nicer.
`"pattern": "^https://ils.rero.ch/schemas/users/*$"`

File: rero_ils/modules/users/jsonschemas/users/user-v0.0.1.json [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+      "pattern": "^\\+[0-9]*$"
+    },
+    "keep_history": {
+      "title": "Keep history",
+      "description": "If enabled, the loan history is saved for a maximum of six months. It is visible to the user and the library staff.",
+      "type": "boolean",
+      "default": false
+    }
+  }
+}
```
Comment (rerowep) [inline rero_ils/modules/users/jsonschemas/users/user-v0.0.1.json:85]: No newline at end of file.

**Review by rerowep — CHANGES_REQUESTED** [review]

File: rero_ils/modules/users/api.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+                value = data.get(field)
+                if value is not None:
+                    if field == 'birth_date':
+                        value = datetime.strptime(value, '%Y-%m-%d')
+                    setattr(profile, field, value)
+            # send the reset password notification for new users
+            if email:
+                user.email = email
+            db.session.merge(user)
+        db.session.commit()
```
[the remark this answers is not in the document]
Comment (BadrAly) [reply rero_ils/modules/users/api.py:75]: `db.session.commit()` is needed.

File: rero_ils/modules/users/api.py [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+                        profile, field,
+                        datetime.strptime(data.get(field), '%Y-%m-%d'))
+                else:
+                    setattr(profile, field, data.get(field, ''))
+            # send reset password if email is changed
+            if email and email != user.email:
+                user.email = email
+                send_reset_password_instructions(user)
+            db.session.merge(user)
+        db.session.commit()
```
[the remark this answers is not in the document]
Comment (BadrAly) [reply rero_ils/modules/users/api.py:105]: `db.session.commit()` is needed.

**Review by iGormilhit — CHANGES_REQUESTED** [review]
Commit message proposition:

```
users: add new resource

* Adds a new 'user' resource to manage the user profiles. It contains
  the user personal data as opposed to the patron data that are link to
  a specific organisation.
* Writes an API to manage user creations and updates.
* Create units testing.
```

File: rero_ils/modules/users/api.py [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+        confirm_user(user)
+        return cls(user)
+
+    def update(self, data):
+        """User record update.
+
+        :param data - dictionary representing a user record to update
+        """
+        self._validate(data=data)
+        email = data.pop('email', None)
```
[the remark this answers is not in the document]
Comment (jma) [reply rero_ils/modules/users/api.py]: Yes the email is on the User database and not on the profile. The schema is only for the profile.

**Review by iGormilhit — CHANGES_REQUESTED** [review]
New commit message proposition:

```
users: add new resource

* Adds a new 'user' resource to manage the user profiles. It contains
  the user personal data as opposed to the patron data that are link to
  a specific organisation.
* Writes a REST API:
  * A GET, POST, PUT methods to create, update and retrieves the patron
    personal information.
  * A search endpoint to retrieve the patron personal information given
    an email or a username.
* Creates units testing.
* Centralizes several views and methods in a specific `User` python
  class.
* Disables the validation email for freshly created user.
* Uses the standard ngx-formly `widget` field for the user JSON schema.
```
Formatting, some minor typos, slighlty reworded.

**Review by lauren-d — APPROVED** [review]

**Review by rerowep — APPROVED** [review]

**Review by iGormilhit — APPROVED** [review]

**Pull request merged**